File: cpp/collatz-conjecture/collatz_conjecture.cpp
#include "collatz_conjecture.h"

namespace collatz_conjecture
{
    int steps(int num)
    {
        if (num < 1)
        {
            throw std::domain_error("Invalid arguments");
        }

        int steps = 0;

        while (num > 1)
        {
            (num % 2 == 0) ? num /= 2 : num = 3 * num + 1;

            steps++;
        }

        return steps;
    }
} // namespace collatz_conjecture

File: ruby/collatz-conjecture/collatz_conjecture.rb
module CollatzConjecture
  def self.steps(num)
    raise ArgumentError if num <= 0

    steps = 0

    until num == 1
      if num.even?
        num /= 2
      else
        num = 3 * num + 1
      end

      steps += 1
    end

    steps
  end
end
